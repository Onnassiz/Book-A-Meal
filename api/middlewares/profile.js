import isUrl from 'is-url';

const { check } = require('express-validator/check');

export const profileFormConstraints = [
  check('businessName')
    .exists()
    .withMessage('the business name field is require')
    .isLength({ min: 1 })
    .withMessage('the business name field is required')
    .isString()
    .withMessage('the business name must be a string')
    .trim(),

  check('mission')
    .exists()
    .withMessage('the business name field is require')
    .isLength({ min: 1 })
    .withMessage('the business name field is required')
    .isString()
    .withMessage('the business name must be a string')
    .trim(),

  check('contact')
    .exists()
    .withMessage('the contact field is require')
    .isLength({ min: 1 })
    .withMessage('the contact field is required')
    .isString()
    .withMessage('the contact must must be a string')
    .trim(),

  check('email')
    .exists()
    .withMessage('the email field is require')
    .isLength({ min: 1 })
    .withMessage('the email field is required')
    .isEmail()
    .withMessage('the email field must be an email')
    .trim()
    .normalizeEmail(),

  check('banner')
    .optional({ nullable: true })
    .custom((value) => {
      return isUrl(value);
    })
    .withMessage('image link must be a URL'),
];

export const profileBannerConstraints = [
  check('banner')
    .exists()
    .withMessage('the image field is require')
    .isLength({ min: 1 })
    .withMessage('the image field is required')
    .optional({ nullable: true })
    .custom((value) => {
      return isUrl(value);
    })
    .withMessage('image link must be a URL'),
];
