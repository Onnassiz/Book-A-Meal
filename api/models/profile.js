
module.exports = (sequelize, DataTypes) => {
  const Profile = sequelize.define('profile', {
    id: {
      allowNull: false,
      primaryKey: true,
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
    },
    businessName: DataTypes.STRING,
    contact: DataTypes.STRING,
    email: DataTypes.STRING,
    mission: DataTypes.STRING,
    banner: DataTypes.STRING,
  }, {});
  Profile.associate = (models) => {
    Profile.belongsTo(models.user);
    // Profile.belongsTo(models.mealOrder);
  };
  return Profile;
};

