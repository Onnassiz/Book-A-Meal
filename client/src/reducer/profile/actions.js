import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actions';

export const SET_PROFILE = 'SET_PROFILE';
export const SET_PROFILE_SERVER_ERRORS = 'SET_PROFILE_SERVER_ERRORS';
export const SET_PROFILE_ALERT = 'SET_PROFILE_ALERT';

function setAlert(alert) {
  return dispatch => dispatch({
    type: SET_PROFILE_ALERT,
    alert,
  });
}

export function setProfile(profile) {
  return dispatch => dispatch({
    type: SET_PROFILE,
    id: profile.id,
    businessName: profile.businessName,
    mission: profile.mission,
    contact: profile.contact,
    banner: profile.banner,
    email: profile.email,
  });
}

function setProfileErrors(errors) {
  return dispatch => dispatch({
    type: SET_PROFILE_SERVER_ERRORS,
    errors,
  });
}

export function postProfile(profile) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.post('api/v1/profile', profile).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Profile successfully created'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function updateProfile(profile) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/profile/${profile.id}`, profile).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Profile successfully updated'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function putImage(id, field) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/profile/image/${id}`, field).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Banner successfully uploaded'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function getProfile() {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.get('api/v1/profile').then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      return error;
    });
  };
}
