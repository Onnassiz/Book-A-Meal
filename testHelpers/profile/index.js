import { profile } from '../../api/models';

import { getCatererId } from '../main';

const deleteProfile = (done) => {
  profile.destroy({
    where: {},
  }).then(() => {
    done();
  });
};

const insertProfile = done => getCatererId(done, false).then((id) => {
  const newProfile = profile.build({
    businessName: 'Your name',
    mission: 'Your mission',
    contact: 'Your this is your contact',
    banner: 'https://banner.com',
    email: 'benjamin@gmail.com',
    userId: id,
  });

  return newProfile.save().then(data => data);
});

export {
  deleteProfile,
  insertProfile,
};
