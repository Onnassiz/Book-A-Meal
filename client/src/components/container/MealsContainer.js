import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { getProfile } from '../../reducer/profile/actions';
import { postMeal, getMeals, updateMeal, deleteMealById } from '../../reducer/meals/actions';
import Meals from '../presentation/Meals';

const mapStateToProps = (state) => {
  return {
    meals: state.meals,
    user: state.user,
    profile: state.profile,
    formState: state.formState,
  };
};


function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    getMeals,
    postMeal,
    getProfile,
    updateMeal,
    deleteMealById,
  }, dispatch);
}


export default withRouter(connect(
  mapStateToProps,
  mapDispatchToProps,
)(Meals));
