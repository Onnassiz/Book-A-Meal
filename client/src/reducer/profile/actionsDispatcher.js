import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actionsDispatcher';
import { setProfileAction, setProfileErrorsAction } from './actions';


export const setProfile = profile => dispatch => dispatch(setProfileAction(profile));

const setProfileErrors = errors => dispatch => dispatch(setProfileErrorsAction(errors));

export const postProfile = profile => (dispatch) => {
  dispatch(setLoading());
  return axios.post('api/v1/profile', profile).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const updateProfile = profile => (dispatch) => {
  dispatch(setLoading());
  return axios.put(`api/v1/profile/${profile.id}`, profile).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const putImage = (id, field) => (dispatch) => {
  dispatch(setLoading());
  return axios.put(`api/v1/profile/${id}`, field).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const getProfile = () => (dispatch) => {
  dispatch(setLoading());
  return axios.get('api/v1/profile').then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};
