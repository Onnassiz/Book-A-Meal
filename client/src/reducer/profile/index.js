import { SET_PROFILE, SET_PROFILE_SERVER_ERRORS } from './actions';

const initialState = {};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_PROFILE:
      return {
        id: action.id,
        businessName: action.businessName,
        email: action.email,
        mission: action.mission,
        contact: action.contact,
        banner: action.banner,
      };
    case SET_PROFILE_SERVER_ERRORS:
      return { ...state, errors: action.errors };
    default:
      return state;
  }
};
