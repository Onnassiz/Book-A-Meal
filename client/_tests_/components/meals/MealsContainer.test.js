import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import MealsContainer from '../../../src/components/container/MealsContainer';
import profile from '../../objectProps/profile.props';

const mockStore = configureStore();

const store = mockStore({
  user: {},
  meals: {},
  profile,
  formState: { isLoading: false },
  postMeal: jest.fn(),
  updateMeal: jest.fn(),
  deleteMealById: jest.fn(),
  getMeals: jest.fn(),
  getProfile: jest.fn(),
});

const wrapper = mount(<MemoryRouter keyLength={0}><MealsContainer store={store} /></MemoryRouter>);

describe('<MealsContainer />', () => {
  it('Should render the Meals container successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
