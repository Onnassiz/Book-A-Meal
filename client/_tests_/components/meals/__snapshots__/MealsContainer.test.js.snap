// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MealsContainer /> Should render the Meals container successfully 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "PUSH",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 1,
        "length": 2,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(Connect(Meals))
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Route
        render={[Function]}
      >
        <Connect(Meals)
          history={
            Object {
              "action": "PUSH",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 1,
              "length": 2,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          store={
            Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            }
          }
        >
          <Meals
            deleteMealById={[Function]}
            formState={
              Object {
                "isLoading": false,
              }
            }
            getMeals={[Function]}
            getProfile={[Function]}
            history={
              Object {
                "action": "PUSH",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 1,
                "length": 2,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            meals={Object {}}
            postMeal={[Function]}
            profile={
              Object {
                "banner": "See me coming",
                "businessName": "The blast",
                "contact": "12 Benny Road Lagos",
                "email": "onnassiz@andela.com",
                "id": "3a3c499f-803a-4e06-a328-307daff22243",
                "mission": "Feeding the world",
              }
            }
            store={
              Object {
                "clearActions": [Function],
                "dispatch": [Function],
                "getActions": [Function],
                "getState": [Function],
                "replaceReducer": [Function],
                "subscribe": [Function],
              }
            }
            storeSubscription={
              Subscription {
                "listeners": Object {
                  "clear": [Function],
                  "get": [Function],
                  "notify": [Function],
                  "subscribe": [Function],
                },
                "onStateChange": [Function],
                "parentSub": undefined,
                "store": Object {
                  "clearActions": [Function],
                  "dispatch": [Function],
                  "getActions": [Function],
                  "getState": [Function],
                  "replaceReducer": [Function],
                  "subscribe": [Function],
                },
                "unsubscribe": [Function],
              }
            }
            updateMeal={[Function]}
            user={Object {}}
          >
            <div
              id="content-body"
            >
              <div
                className="col-12"
              >
                <button
                  className="button"
                  id="addMeal"
                  onClick={[Function]}
                >
                  <i
                    className="ion-android-add"
                  />
                   Add Meal
                </button>
                <div>
                  <div />
                  <div>
                    <MealModal
                      addToMeals={[Function]}
                      formState={
                        Object {
                          "isLoading": false,
                        }
                      }
                      handleAddButtonClick={[Function]}
                      isShowingModal={false}
                      meals={Object {}}
                      postMeal={[Function]}
                      state={
                        Object {
                          "activePage": 1,
                          "currentMeal": Object {},
                          "isShowingAddPhoto": false,
                          "isShowingDropZoneModal": false,
                          "isShowingModal": false,
                          "meals": Array [],
                          "mealsCount": 0,
                          "updateMode": false,
                        }
                      }
                      updateComponentMeals={[Function]}
                      updateMeal={[Function]}
                    >
                      <div>
                        <Modal
                          ariaHideApp={false}
                          bodyOpenClassName="ReactModal__Body--open"
                          className="create-profile"
                          closeTimeoutMS={1}
                          contentLabel="Modal"
                          isOpen={false}
                          onAfterOpen={[Function]}
                          onRequestClose={[Function]}
                          parentSelector={[Function]}
                          portalClassName="ReactModalPortal"
                          shouldCloseOnEsc={true}
                          shouldCloseOnOverlayClick={true}
                          shouldFocusAfterRender={true}
                          shouldReturnFocusAfterClose={true}
                          style={
                            Object {
                              "content": Object {
                                "WebkitOverflowScrolling": "touch",
                                "background": "#ffffff",
                                "borderRadius": "6px",
                                "bottom": "15%",
                                "left": "33%",
                                "outline": "none",
                                "overflow": "auto",
                                "padding": "12px",
                                "position": "absolute",
                                "right": "34%",
                                "top": "10%",
                              },
                              "overlay": Object {
                                "backgroundColor": "rgba(0, 0, 0, 0.4)",
                                "bottom": 0,
                                "left": 0,
                                "position": "fixed",
                                "right": 0,
                                "top": 0,
                              },
                            }
                          }
                        >
                          <Portal
                            containerInfo={
                              <div
                                class="ReactModalPortal"
                              />
                            }
                          >
                            <ModalPortal
                              ariaHideApp={false}
                              bodyOpenClassName="ReactModal__Body--open"
                              className="create-profile"
                              closeTimeoutMS={1}
                              contentLabel="Modal"
                              defaultStyles={
                                Object {
                                  "content": Object {
                                    "WebkitOverflowScrolling": "touch",
                                    "background": "#fff",
                                    "border": "1px solid #ccc",
                                    "borderRadius": "4px",
                                    "bottom": "40px",
                                    "left": "40px",
                                    "outline": "none",
                                    "overflow": "auto",
                                    "padding": "20px",
                                    "position": "absolute",
                                    "right": "40px",
                                    "top": "40px",
                                  },
                                  "overlay": Object {
                                    "backgroundColor": "rgba(255, 255, 255, 0.75)",
                                    "bottom": 0,
                                    "left": 0,
                                    "position": "fixed",
                                    "right": 0,
                                    "top": 0,
                                  },
                                }
                              }
                              isOpen={false}
                              onAfterOpen={[Function]}
                              onRequestClose={[Function]}
                              parentSelector={[Function]}
                              portalClassName="ReactModalPortal"
                              shouldCloseOnEsc={true}
                              shouldCloseOnOverlayClick={true}
                              shouldFocusAfterRender={true}
                              shouldReturnFocusAfterClose={true}
                              style={
                                Object {
                                  "content": Object {
                                    "WebkitOverflowScrolling": "touch",
                                    "background": "#ffffff",
                                    "borderRadius": "6px",
                                    "bottom": "15%",
                                    "left": "33%",
                                    "outline": "none",
                                    "overflow": "auto",
                                    "padding": "12px",
                                    "position": "absolute",
                                    "right": "34%",
                                    "top": "10%",
                                  },
                                  "overlay": Object {
                                    "backgroundColor": "rgba(0, 0, 0, 0.4)",
                                    "bottom": 0,
                                    "left": 0,
                                    "position": "fixed",
                                    "right": 0,
                                    "top": 0,
                                  },
                                }
                              }
                            />
                          </Portal>
                        </Modal>
                      </div>
                    </MealModal>
                    <MealsTable
                      state={
                        Object {
                          "activePage": 1,
                          "currentMeal": Object {},
                          "isShowingAddPhoto": false,
                          "isShowingDropZoneModal": false,
                          "isShowingModal": false,
                          "meals": Array [],
                          "mealsCount": 0,
                          "updateMode": false,
                        }
                      }
                      toggleAddPhoto={[Function]}
                      toggleShowDeleteModal={[Function]}
                      toggleUpdateModal={[Function]}
                    >
                      <div>
                        <div
                          id="no-menu"
                        >
                          <h2>
                            You do not have any meals
                          </h2>
                        </div>
                      </div>
                    </MealsTable>
                  </div>
                  <div />
                  <ImageUploader
                    isShowingAddPhoto={false}
                    putImage={[Function]}
                    toggleAddPhoto={[Function]}
                  >
                    <div
                      className="col-12"
                    >
                      <Modal
                        ariaHideApp={false}
                        bodyOpenClassName="ReactModal__Body--open"
                        className="image-upload"
                        closeTimeoutMS={1}
                        contentLabel="Modal"
                        isOpen={false}
                        parentSelector={[Function]}
                        portalClassName="ReactModalPortal"
                        shouldCloseOnEsc={true}
                        shouldCloseOnOverlayClick={true}
                        shouldFocusAfterRender={true}
                        shouldReturnFocusAfterClose={true}
                        style={
                          Object {
                            "content": Object {
                              "WebkitOverflowScrolling": "touch",
                              "background": "#ffffff",
                              "border": "none",
                              "borderRadius": "6px",
                              "bottom": "30%",
                              "left": "33%",
                              "outline": "none",
                              "overflow": "auto",
                              "padding": "15px",
                              "position": "absolute",
                              "right": "33%",
                              "top": "15%",
                            },
                            "overlay": Object {
                              "backgroundColor": "rgba(0, 0, 0, 0.4)",
                              "bottom": 0,
                              "left": 0,
                              "position": "fixed",
                              "right": 0,
                              "top": 0,
                            },
                          }
                        }
                      >
                        <Portal
                          containerInfo={
                            <div
                              class="ReactModalPortal"
                            />
                          }
                        >
                          <ModalPortal
                            ariaHideApp={false}
                            bodyOpenClassName="ReactModal__Body--open"
                            className="image-upload"
                            closeTimeoutMS={1}
                            contentLabel="Modal"
                            defaultStyles={
                              Object {
                                "content": Object {
                                  "WebkitOverflowScrolling": "touch",
                                  "background": "#fff",
                                  "border": "1px solid #ccc",
                                  "borderRadius": "4px",
                                  "bottom": "40px",
                                  "left": "40px",
                                  "outline": "none",
                                  "overflow": "auto",
                                  "padding": "20px",
                                  "position": "absolute",
                                  "right": "40px",
                                  "top": "40px",
                                },
                                "overlay": Object {
                                  "backgroundColor": "rgba(255, 255, 255, 0.75)",
                                  "bottom": 0,
                                  "left": 0,
                                  "position": "fixed",
                                  "right": 0,
                                  "top": 0,
                                },
                              }
                            }
                            isOpen={false}
                            parentSelector={[Function]}
                            portalClassName="ReactModalPortal"
                            shouldCloseOnEsc={true}
                            shouldCloseOnOverlayClick={true}
                            shouldFocusAfterRender={true}
                            shouldReturnFocusAfterClose={true}
                            style={
                              Object {
                                "content": Object {
                                  "WebkitOverflowScrolling": "touch",
                                  "background": "#ffffff",
                                  "border": "none",
                                  "borderRadius": "6px",
                                  "bottom": "30%",
                                  "left": "33%",
                                  "outline": "none",
                                  "overflow": "auto",
                                  "padding": "15px",
                                  "position": "absolute",
                                  "right": "33%",
                                  "top": "15%",
                                },
                                "overlay": Object {
                                  "backgroundColor": "rgba(0, 0, 0, 0.4)",
                                  "bottom": 0,
                                  "left": 0,
                                  "position": "fixed",
                                  "right": 0,
                                  "top": 0,
                                },
                              }
                            }
                          />
                        </Portal>
                      </Modal>
                    </div>
                  </ImageUploader>
                </div>
              </div>
            </div>
          </Meals>
        </Connect(Meals)>
      </Route>
    </withRouter(Connect(Meals))>
  </Router>
</MemoryRouter>
`;
